### ISSUE TITLE
There is no way to search through filtered applications/software that isn't installed without providing a search query + bugged filters?

### ISSUE THREAD

**Issue by Templayer** [vinifmor/bauh#245]
The filters do not seem to allow me to search through software that hasn't been installed if I don't type anything in the search input.

Example:

![image](https://user-images.githubusercontent.com/11728067/159894152-7c300c0e-e33d-4ddf-a989-8c561a15396f.png)

I wish to get a list of all available compression tools in bauh. I don't know the name of the compression tool that I'm going to need, as that is why I need them listed,  so I cannot put anything in the search input. Toggling "App" doesn't do anything in this case.

Also, I tried inputting just whitespace into the search input, but that didn't work.

And I think I just managed to bork it.

First, I tried using "-" in the search input, which did partially get me the result I needed, but the filters were unset.

![image](https://user-images.githubusercontent.com/11728067/159895612-cc3a134b-c990-410a-9381-cb8c2ccf2875.png)

So I've tried setting the filter to compression software again.

![image](https://user-images.githubusercontent.com/11728067/159895303-cf46e163-059d-49fe-b898-926b629652b4.png)

The Installed button is there, the Installed filter checkbox is there and it is unchecked, yet it contains only installed software!

**Comment by Templayer** [comment]
PS: Unless there is no Compression tagged software that has a "-" in its package name that I don't have installed. Which I don't think is the case.

**Comment by vinifmor** [comment]
@Templayer , all the search is provided by the packaging technology back-ends (flatpak, aptitude, ...). So, if nothing is found when your search for "-", is because they did not find matching packages.

About filtering non-installed packages: it is working perfectly on my end (category, name, type, installed)

**Comment by Templayer** [comment]
@vinifmor I don't think we understand each other.

Searching for "-" DOES find all sorts of applications that are not installed. As shown in my printscreens (please look at them again!). But searching for "-" removes any other filter. If I THEN set another filter (like a category), it lists all the installed applications ONLY, even though the "Installed" checkbox is unchecked!

> About filtering non-installed packages: it is working perfectly on my end (category, name, type, installed)

Very well. Please provide a walkthrough on how to make bauh list ALL Compression software, that is not installed. The "Installed" filter and button are not available if the search input is not given.

![image](https://user-images.githubusercontent.com/11728067/159936888-a3f08f9c-b0c2-4834-a2e2-fb898b81173a.png)

What should I do next? What should I click on the printscreen given? What should I input? Elaborate, please.

**Comment by Templayer** [comment]
@vinifmor After updating to a new unreleased version ( https://github.com/vinifmor/bauh/issues/242 ) , the filters are now working correctly!

![image](https://user-images.githubusercontent.com/11728067/160079816-7c77ec2f-f710-4c16-a97f-99606bd1cfdf.png)

OK, I spoke too soon. I've managed to bork them again. Looks like there is a specific trigger condition.

Found it - the filter "Category" is bugged - while the Application/Software type filter (Flatpack etc.) works as intended (presumably - filtering either both installed and not installed software if the Installed checkbox filter is unchecked or only installed ones if it is checked), the "Category" filter always acts as if the Installed checkbox filter is checked even if it isn't.

I can record a video if you don't believe or understand me.

**Comment by vinifmor** [comment]
The "Installed" and "Apps" checkboxes don't ignore their opposite meaning. Their function is to provide a quick filtering for their meaning. Is that the confusion ? If so, I can add some tooltips.

**Comment by Templayer** [comment]
My point is that the App Type filter (flatpack for example) and the Category filter act differently!

If App Type is defined in the filter, then it shows both installed and not installed (without checking the Installed checkbox).
If Category is defined in the filter, then it ONLY shows installed (no matter what the Installed checkbox state is).

They need to act the same way.

I'm going to have to record a video during the weekend... to show you what I mean.

Basically, it is currently absolutely and completely impossible to use the Category filter AND see non-installed apps in the list.

Please make me a printscreen where you have the category filter used (a category is selected there) and the list shows apps that are not installed. Please. Please.

**Comment by vinifmor** [comment]
Perhaps the software you are looking for is not categorized ? Not all of them are. 

Here is an example:

1) Searching for "database" (displaying only **50/864**):
![bauh_search](https://user-images.githubusercontent.com/4312414/160116602-23e476d5-54f6-4cd0-ae9f-a29c1649b280.png)


 2) Applying the "Development" category filter (displaying both installed and not installed software as expected -> **12/864**):
![bauh_search_dev_category](https://user-images.githubusercontent.com/4312414/160116190-539e06ee-fdf8-41cc-bd6c-4362fbadbdee.png)

2) Applying the "Installed" filter (only DBeaver was displayed -> **1/864**):
![bauh_search_dev_category_installed](https://user-images.githubusercontent.com/4312414/160116379-615dd9d9-fbf0-460e-820b-0e163d664dd3.png)

**Comment by vinifmor** [comment]
Here is an example of a search that didn't return categorized software (the category filter is not even displayed)
![bauh_search_non_categorized](https://user-images.githubusercontent.com/4312414/160119949-5c072e0d-3710-4cf1-a02f-0a50f545adc7.png)

Probably in your case some of the software returned were categorized, but the Flatpaks returned were not.

**Comment by Templayer** [comment]
You are completely correct. And I was completely wrong. I wasn't aware that some applications weren't categorized. (and I've just checked all the comments in this Issue and it seems like I wasn't told that, nor does the application say that anywhere)

Maybe a Category column is in order? It is a little bit silly to have a filter for something that the list doesn't even show. :P  

Also, there is still that issue of not being able to "browse" all applications (installed and non-installed) without using that fulltext input (supertux in your latest printscreen). Just checked the comments here and I think you never answered that. It might just be that I'm missing some kind of a catch.


While we are at it, why is development twice in the filter?

![image](https://user-images.githubusercontent.com/11728067/160121090-41c83b6a-0f8f-4d84-aa20-db4c74f13a22.png)

And Network? (and there are going to be more of them, presumably. I think I also saw email twice, but in different spots)

![image](https://user-images.githubusercontent.com/11728067/160121449-0d02fead-83a1-4b31-b5e3-4ef50c688edb.png)

Is it because two app types use the same name for a category? If that is the case, shouldn't those get consolidated?

**Comment by vinifmor** [comment]
Just to clarify: the upper "Search" bar asks the back-ends to perform a search based on the words you typed. The filters are used to perform sub-searches within the returned results (the back-ends do not interfere, just the UI). You would have to use the "name" filter for your use case.

About the categories: they should have been merged. It is probably a separate issue.

Another one: the "Installed" filter should not be displayed f there are no installed applications displayed (as you mentioned)

**Comment by Templayer** [comment]
> Just to clarify: the upper "Search" bar asks the back-ends to perform a search based on the words you typed.

And do they support searching for nothing, i.e. browsing everything?

**Comment by vinifmor** [comment]
Sorry for the late reply. It is not possible to search for an empty String. Some back-ends don't support this (probably because thousands of packages would be returned), so bauh does not even proceed with the call.

**Comment by Templayer** [comment]
And there isn't a special way to get all packages? For example, those that would support searching for an empty string would use that, while for those that don't support it a different method to obtain all packages would be used instead.

It just feels weird to me that the packaging mechanisms don't have some kind of a master index that could be browsed. Not common to each other, of course. But each of them individually should have something like that. :-S

Let me try to put my two cents into this...

Debian supports it, I was able to get a list of all packages from there. While specifying what I presume is the debian version codename: https://packages.debian.org/buster/allpackages

Flatpacks seem to be browsable to "all" - https://flathub.org/apps/category/All

Aur seems to be doable at a first glance - https://bbs.archlinux.org/viewtopic.php?pid=1978090#p1978090

After some fiddling, I think I might have managed to list all snaps in a rather hilarious manner. By using two apostrophes '. Tried quotation marks at first, but those didn't work.
https://snapcraft.io/search?q=%27%27

AppImageHub can browse all AppImages - https://www.appimagehub.com/browse

**Comment by vinifmor** [comment]
bauh doesn't use these interfaces for querying. Most of the extensions are associated with an official back-end application.

**Comment by Templayer** [comment]
Maybe in this specific instance it should. :^) 

But it is a lot of work. Maybe you could just ask them to provide methods in their back-end applications for this and see if they respond?
